doc/guides/kconfig/tips.rst

```diff
@@ -416,6 +420,40 @@ interface, to make sure that things behave the way you expect. This is
 especially true when making moderately complex changes like these.
 
 
+Assignments to promptless symbols in configuration files
+********************************************************
+
+Assignments to invisible (promptless) symbols in configuration files are always
+ignored. Invisible symbols get their value indirectly from other symbols, via
+e.g. ``default`` and ``select``.
+
+A common source of confusion is opening the output configuration file
+(:file:`zephyr/.config`), seeing a bunch of assignments to invisible symbols,
+and assuming that those assignments must be respected when the configuration is
+read back in by Kconfig. In reality, all assignments to invisible symbols in
+:file:`zephyr/.config` are ignored by Kconfig, like for other configuration
+files.
+
+To understand why :file:`zephyr/.config` still includes assignments to
+invisible symbols, it helps to realize that :file:`zephyr/.config` serves two
+separate purposes:
+
+1. It holds the saved configuration, and
+
+2. it holds configuration output. :file:`zephyr/.config` is parsed by the CMake
+   files to let them query configuration settings, for example.
+
+The assignments to invisible symbols in :file:`zephyr/.config` are just
+configuration output. Kconfig itself ignores assignments to invisible symbols
+when calculating symbol values.
+
+.. note::
```

Comment:
`olddefconfig` isn't used by Zephyr, so wonder if it's be confusing. :thinking:

Reply:
Why is `olddefconfig` removed by Zephyr? It doesn't seem like it would hurt...

I haven't look at it in detail, but again I feel like most of the extremely useful Kconfig documentation hosted by Zephyr is not specific to Zephyr and not getting the larger https://github.com/ulfalizer/Kconfiglib visibility (!) it deserves.

BTW I just quickly hacked a new `genconfig.py --sane` option, it was much easier than I thought it would be. In case anyone's interested:

```diff
--- a/Kconfiglib-subrepo/genconfig.py
+++ b/Kconfiglib-subrepo/genconfig.py
@@ -68,6 +68,13 @@ deps/auto.conf instead. --config-out is meant for cases where incremental build
 information isn't needed.
 """)
 
+    parser.add_argument(
+        "--sane", action='store_true',
+        help="""
+Requires a --config-out parameter. Compares .config input and config-out output
+and fail if they differ.
+""") 
+
     parser.add_argument(
         "--sync-deps",
         metavar="OUTPUT_DIR",
@@ -114,11 +121,17 @@ only supported for backwards compatibility).
     kconf = kconfiglib.Kconfig(args.kconfig_filename)
     print(kconf.load_config(args.dot_config))
 
-    kconf.write_autoconf(args.header_path)
-
     if args.config_out is not None:
         kconf.write_config(args.config_out, save_old=False)
 
+    if args.sane:
+        if not kconf._contents_eq(args.dot_config, kconf._config_contents()):
+            sys.exit("{} and {} are different, try olddefconfig".format(
+                args.dot_config, args.config_out)
+            )
+
+    kconf.write_autoconf(args.header_path)
+
     if args.sync_deps is not None:
         kconf.sync_deps(args.sync_deps)
 
diff --git a/Kconfiglib-subrepo/kconfiglib.py b/Kconfiglib-subrepo/kconfiglib.py
index 37ba501afcb2..6a9680f487c1 100644
--- a/Kconfiglib-subrepo/kconfiglib.py
+++ b/Kconfiglib-subrepo/kconfiglib.py
@@ -1415,8 +1416,10 @@ class Kconfig(object):
 
         return "".join(chunks)
 
+    HEADER = "# Generated by Kconfiglib (https://github.com/ulfalizer/Kconfiglib)\n"
+
     def write_config(self, filename=None,
-                     header="# Generated by Kconfiglib (https://github.com/ulfalizer/Kconfiglib)\n",
+                     header=HEADER,
                      save_old=True, verbose=None):
         r"""
         Writes out symbol values in the .config format. The format matches the
@@ -1491,7 +1494,7 @@ class Kconfig(object):
 
         return "Configuration saved to '{}'".format(filename)
 
-    def _config_contents(self, header):
+    def _config_contents(self, header=HEADER):
         # write_config() helper. Returns the contents to write as a string,
         # with 'header' at the beginning.
         #
```

Reply:
> Why is `olddefconfig` removed by Zephyr? It doesn't seem like it would hurt...

Yeah... could maybe import it along with some other scripts, though things like `all{yes,no}config.py` is probably broken for Zephyr.

Note that the header will still be "as-if" you had run `olddefconfig`, because Kconfig never writes output that doesn't respect dependencies. It can be handy if you inspect `.config` manually though.

Re. `--sane`, I think many people rely on not having to run `olddefconfig`, and update the configuration file via e.g. `menuconfig` instead. Might be a bit too special-cased (though I know it wasn't meant as a PR :).

Reply:
> Re. --sane,  I think many people rely on not having to run olddefconfig, and update the configuration file via e.g. menuconfig instead. 

Two different tools for two different jobs. `menuconfig` is best for learning and exploration. `olddefconfig` is when you know exactly the one option you want to change and quickly `[git] diff` and see all the consequences. In that case it's much faster to use your favorite editor.

`olddefconfig` can also be scripted, `menuconfig` obviously not https://www.chromium.org/chromium-os/how-tos-and-troubleshooting/kernel-configuration#TOC-Generate-a-new-splitconfig-after-adding-new-Kconfig-options

Command line or web browser? Both of course.

> (though I know it wasn't meant as a PR :).

Indeed it wasn't, however the more my build system uses it and the more I love it.